opm/material/fluidsystems/GenericFluidSystem.hh

```diff
@@ -0,0 +1,215 @@
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+
+  Consult the COPYING file in the top-level source directory of this
+  module for the precise wording of the license and the list of
+  copyright holders.
+*/
+
+#ifndef OPM_GENERICFLUIDSYSTEM_HH
+#define OPM_GENERICFLUIDSYSTEM_HH
+
+#include <opm/material/fluidsystems/BaseFluidSystem.hpp>
+
+// TODO: this is something else need to check
+#include <opm/material/fluidsystems/PTFlashParameterCache.hpp>
+#include <opm/material/viscositymodels/LBC.hpp>
+
+namespace Opm {
+
+    template <typename Scalar>
+    struct ComponentParam {
+        std::string name;
+        Scalar molar_mass;
+        Scalar critic_temp;
+        Scalar critic_pres;
+        Scalar critic_vol;
+        Scalar acentric_factor;
+
+        ComponentParam(const std::string_view name_, const Scalar molar_mass_, const Scalar critic_temp_,
+                       const Scalar critic_pres_, const Scalar critic_vol_, const Scalar acentric_factor_)
+                       : name(name_),
+                         molar_mass(molar_mass_),
+                         critic_temp(critic_temp_),
+                         critic_pres(critic_pres_),
+                         critic_vol(critic_vol_),
+                         acentric_factor(acentric_factor_)
+        {}
+    };
+/*!
+ * \ingroup FluidSystem
+ *
+ * \brief A two phase three component fluid system with components
+ * CO2, Methane and NDekan
+ */
+    template<class Scalar, int NumComp>
+    class GenericFluidSystem : public Opm::BaseFluidSystem<Scalar, GenericFluidSystem<Scalar, NumComp> > {
+    public:
+        // TODO: I do not think these should be constant in fluidsystem, will try to make it non-constant later
+        static const int numPhases=2;
+        static const int numComponents = NumComp;
+        static const int numMisciblePhases=2;
+        static const int numMiscibleComponents = 3;
+        // TODO: phase location should be more general
+        static constexpr int oilPhaseIdx = 0;
+        static constexpr int gasPhaseIdx = 1;
+
+        template <class ValueType>
+        using ParameterCache = Opm::PTFlashParameterCache<ValueType, GenericFluidSystem<Scalar, NumComp>>;
+        using ViscosityModel = typename Opm::ViscosityModels<Scalar, GenericFluidSystem<Scalar, NumComp>>;
+        using PengRobinsonMixture = typename Opm::PengRobinsonMixture<Scalar, GenericFluidSystem<Scalar, NumComp>>;
+
+        template<typename Param>
+        static void addComponent(const Param& param)
+        {
+            assert(component_param_.size() <= numComponents);
```

Comment:
You need to `#include <cassert>` to bring in a declaration of the `assert()` macro.